### DISCUSSION TITLE
How can I expose functions to the engine?

### DISCUSSION THREAD

**Discussion by furesoft** [leppie/IronScheme#92] [Discussions]
I want to integrate IronScheme and want to know how to add custom functions and objects to the engine

**Comment by leppie** [comment]
Your question is a little vague. Can you give an example?

**Comment by furesoft** [comment]
I want to use ironscheme as script engine. so i need to add application objects and some functions to the engine to evaluate it

**Comment by leppie** [comment]
Do you just want to call .NET methods? Just use `clr-call` then. 

Plenty of examples in the ironscheme libs. Example:
https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/lib/ironscheme/datetime.sls

**Reply by furesoft** [reply]
No I want to expose instances to the engine

**Reply by leppie** [reply]
I want to help, but I dont know what you really want. Can you give me pseudo code of what you want to achieve from the IronScheme side?

**Comment by bcalco** [comment]
I think I know where he's coming from. The question I have that seems related is, what are the ways an application can explicitly extend Scheme with forms that map to something in the application namespace? How can it call Scheme code that consumes types in the app namespace?

The IronScheme built-ins give a clue... some annotations on static class methods, and the extension method for delegates. But there really isn't an example of an app that explicitly exposes, say, events that can be scripted in IronScheme. It could be desirable for example to have an event such as "onApplicationLoad" which should take a Callable that allows the app to pass in some context parameters, so that, if a handler exists, it can be called from the host on completion of application loading. This is a scenario I'm trying to figure out right now. If I figure out the idiomatic way to do this, I'll share my learning. But it would save time to have some examples, if possible. :)

**Reply by bcalco** [reply]
I find that there is no problem opening a namespace in the app and calling objects from the app; they seem to already be there and are callable. What is not working is the use of the `[Builtin(...)]` attribute ... how does the detection of app-defined built-ins work from the IronScheme side at runtime?

As an example, I am able to call the following static method defined in F# (created in a CLR compatible way):

```fsharp
namespace MyApp.MyClassNamespace 

[<Class>]
type MyStaticClass =

  [<Builtin("myapp-say-hello")>]
  static member SayHello() =
    Console.WriteLine ("Hello")
```

like this:

```
> (import (ironscheme clr))
> (clr-using MyApp.MyClassNamespace)
> (clr-static-call MyStaticClass SayHello)
Hello
> (myapp-say-hello)
Unhandled exception during evaluation:
&undefined
&message: "attempted to use undefined symbol"
&irritants: (myapp-say-hello)
```
Is the detection of `[<Builtin(...)>]` limited to the IronScheme assembly's definitions?

What I'm more interested in, is being able to establish host event handlers that I can optionally set on the Scheme side, such that the host can check if such a handler is set in current/interactive environment, and if so, call it when the event needs to be fired. Some of the parameters will be types defined in the app--it's on the implementer of a handler on the scheme side to know how to process them using normal clr interop methods.

**Reply by leppie** [reply]
I dont think I took `Builtin` further than providing support for internal functionality. IIRC it was a leftover from IronLisp :D 

Simple forms event handler example (12 year old code...): https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/examples/forms.sls

**Reply by bcalco** [reply]
I remember now I had seen that and it helps... however I'm also wondering if it's possible on the host side to look for a form in the current/interactive environment, and call it specifically (if it exists). I feel like I saw an example of that somewhere but rummaging through the wiki I seem to be missing it. Maybe it was in one of these discussion threads... I'll poke around here again.

**Reply by bcalco** [reply]
BTW, for @furesoft's benefit, I'll point out that given the clr calls in `(ironscheme clr)` & friends it's at least possible to reach into the app's namespace and write custom wrappers in libraries added to the `/lib` dir like so:

<install-dir>/lib/myapp/runtime.sls:
```scheme
(library (myapp runtime)
  (export say-hello)
  (import (ironscheme)
          (ironscheme clr))

  (clr-using MyApp.MyClassNamespace)

  (define say-hello
    (lambda ()
      (clr-static-call MyStaticClass SayHello)))
)
```

then from the REPL for example:

```scheme
> (import (myapp runtime))
> (say-hello)
Hello
>
```

**Reply by bcalco** [reply]
@leppie Some of the clues I'm looking for are from your final answers to [this thread](https://github.com/IronScheme/IronScheme/discussions/82).
